code/modules/food_and_drinks/drinks/drinks.dm

```diff
@@ -270,6 +270,40 @@
 	materials = list(MAT_METAL=1500)
 	amount_per_transfer_from_this = 10
 	volume = 100
+	var/shaking = FALSE
+
+/obj/item/reagent_containers/food/drinks/shaker/Initialize(mapload)
+	. = ..()
+	reagents.set_reacting(FALSE)
+
+/obj/item/reagent_containers/food/drinks/shaker/attack_self(mob/user)
+	if(!reagents.total_volume)
+		to_chat(user, "<span class='warning'>You won't shake an empty shaker now, will you?</span>")
+		return
+
+	if(shaking)
+		return
+
+	shaking = TRUE
+	var/adjective = pick("furiously", "passionately", "with vigor", "with determination", "like a devil", "with care and love", "like there is no tomorrow")
+	user.visible_message("<span class='notice'>\The [user] shakes \the [src] [adjective]!</span>", "<span class='notice'>You shake \the [src] [adjective]!</span>")
+	icon_state = "shaker-shake"
+	if(iscarbon(loc))
```

Comment:
you perform an `iscarbon()` check on `loc` twice here, may want to just do this check once and store it to a variable. Additionally, any reason why you're using `loc` here instead of just type checking the user param?